src/fibsem_tools/io/h5.py

```diff
@@ -7,78 +7,116 @@
 
 H5_ACCESS_MODES = ("r", "r+", "w", "w-", "x", "a")
 
-H5_DATASET_KWDS = ("name",
-                     "shape",
-                     "dtype",
-                     "data",
-                     "chunks",
-                     "compression",
-                     "compression_opts",
-                     "scaleoffset",
-                     "shuffle",
-                     "fletcher32",
-                     "maxshape",
-                     "fillvalue",
-                     "track_times",
-                     "track_order",
-                     "external",
-                     "allow_unknown_filter")
-
-H5_GROUP_KWDS = ("name",
-                "track_order")
-
-H5_FILE_KWDS = ("name",
-                  "mode",
-                  "driver",
-                  "libver",
-                  "userblock_size",
-                  "swmr",
-                  "rdcc_nslots",
-                  "rdcc_nbytes",
-                  "rdcc_w0",
-                  "track_order",
-                  "fs_strategy",
-                  "fs_persist",
-                  "fs_threshold")
+H5_DATASET_KWDS = (
+    "name",
+    "shape",
+    "dtype",
+    "data",
+    "chunks",
+    "compression",
+    "compression_opts",
+    "scaleoffset",
+    "shuffle",
+    "fletcher32",
+    "maxshape",
+    "fillvalue",
+    "track_times",
+    "track_order",
+    "dcpl",
+    "external",
+    "allow_unknown_filter",
+)
+
+H5_GROUP_KWDS = ("name", "track_order")
+
+H5_FILE_KWDS = (
+    "name",
+    "mode",
+    "driver",
+    "libver",
+    "userblock_size",
+    "swmr",
+    "rdcc_nslots",
+    "rdcc_nbytes",
+    "rdcc_w0",
+    "track_order",
+    "fs_strategy",
+    "fs_persist",
+    "fs_threshold",
+)
+
+
+# Could use multiple inheritance here
+class ManagedDataset(h5py.Dataset):
+    """
+    h5py.Dataset with context manager behavior
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        self.file.close()
+
+
+class ManagedGroup(h5py.Group):
+    """
+    h5py.Group with context manager behavior
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, ex_type, ex_value, ex_traceback):
+        self.file.close()
+
 
 def partition_h5_kwargs(**kwargs) -> Tuple[Dict[str, Any], Dict[str, Any]]:
     """
     partition kwargs into file-creation kwargs and dataset-creation kwargs
     """
     file_kwargs = kwargs.copy()
     dataset_kwargs = {}
-    for key in H5_DATASET_KWDS:
-        if key in file_kwargs:
+    for key in kwargs:
+        if key in H5_DATASET_KWDS:
             dataset_kwargs[key] = file_kwargs.pop(key)
 
     return file_kwargs, dataset_kwargs
 
 
 def access_h5(
-    store: Pathlike, path: Pathlike, mode: str, **kwargs
+    store: Union[h5py.File, Pathlike], path: Pathlike, **kwargs
```

Comment:
```suggestion
    store: Union[h5py.File, Pathlike], path: Union[Pathlike, NoneType], **kwargs
```

`None` is a valid dataset name and is used to create anonymous datasets.

Reply:
`NoneType` returned in 3.10 https://docs.python.org/3/library/types.html?highlight=nonetype#types.NoneType

Reply:
interesting, how is that different from naming a dataset with the empty string?

Reply:
For one, providing an empty string results in a `TypeError` in h5py:

```python
In [22]: with h5py.File("test.hdf5", "w") as h5f:
    ...:     h5f.create_dataset(None, data = np.zeros((5,5)))
    ...:     print(list(h5f.keys()))
    ...:
    ...:
[]

In [23]: with h5py.File("test.hdf5", "w") as h5f:
    ...:     h5f.create_dataset("", data = np.zeros((5,5)))
    ...:     print(list(h5f.keys()))
    ...:
    ...:
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-23-5230edf15b52> in <module>
      1 with h5py.File("test.hdf5", "w") as h5f:
----> 2     h5f.create_dataset("", data = np.zeros((5,5)))
      3     print(list(h5f.keys()))
      4
      5

~\.julia\conda\3\lib\site-packages\h5py\_hl\group.py in create_dataset(self, name, shape, dtype, data, **kwds)
    147                     group = self.require_group(parent_path)
    148
--> 149             dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
    150             dset = dataset.Dataset(dsid)
    151             return dset

~\.julia\conda\3\lib\site-packages\h5py\_hl\dataset.py in make_new_dset(parent, shape, dtype, data, name, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times, external, track_order, dcpl, allow_unknown_filter)
    140
    141
--> 142     dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl)
    143
    144     if (data is not None) and (not isinstance(data, Empty)):

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\h5d.pyx in h5py.h5d.create()

TypeError: expected bytes, str found

In [24]: with h5py.File("test.hdf5", "w") as h5f:
    ...:     h5f.create_dataset(b"", data = np.zeros((5,5)))
    ...:     print(list(h5f.keys()))
    ...:
    ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-24-7f1869f34ebe> in <module>
      1 with h5py.File("test.hdf5", "w") as h5f:
----> 2     h5f.create_dataset(b"", data = np.zeros((5,5)))
      3     print(list(h5f.keys()))
      4
      5

~\.julia\conda\3\lib\site-packages\h5py\_hl\group.py in create_dataset(self, name, shape, dtype, data, **kwds)
    147                     group = self.require_group(parent_path)
    148
--> 149             dsid = dataset.make_new_dset(group, shape, dtype, data, name, **kwds)
    150             dset = dataset.Dataset(dsid)
    151             return dset

~\.julia\conda\3\lib\site-packages\h5py\_hl\dataset.py in make_new_dset(parent, shape, dtype, data, name, chunks, compression, shuffle, fletcher32, maxshape, compression_opts, fillvalue, scaleoffset, track_times, external, track_order, dcpl, allow_unknown_filter)
    140
    141
--> 142     dset_id = h5d.create(parent.id, name, tid, sid, dcpl=dcpl)
    143
    144     if (data is not None) and (not isinstance(data, Empty)):

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\_objects.pyx in h5py._objects.with_phil.wrapper()

h5py\h5d.pyx in h5py.h5d.create()

ValueError: Unable to create dataset (no name given)
```